crates/wasmparser/src/validator/component_types.rs

```diff
@@ -2496,7 +2505,13 @@ macro_rules! limits_match {
 
 impl<'a> SubtypeCx<'a> {
     /// Create a new instance with the specified type lists
+    ///
+    /// # Panics
+    ///
+    /// Panics if the two given `TypesRef`s are not associated with the same
+    /// `Validator`.
     pub fn new_with_refs(a: TypesRef<'a>, b: TypesRef<'a>) -> SubtypeCx<'a> {
+        assert_eq!(a.id(), b.id());
         Self::new(a.list, b.list)
     }
```

Comment:
Could this go further and take a single `TypesRef` as a constructor argument?

Reply:
I tried that, but it doesn't work because while the `TypesRef`s are always associated with the same validator, they can still be distinct and have different sets of component types. See my last paragraph in the OP:

> I had originally hoped to move the `SubtypeCx` to wrapping a single `TypesRef`, but this cannot work because (unlike core types) new component model types can be pushed into each different `TypesRef` for different components.

Reply:
Wait hang on, while `TypesRef` can be different aren't the`TypeList` values the same? Otherwise how are core type canonicalizations shared?

Reply:
The underlying `TypeList`s are not necessarily the same, but one will always be the extension of another, so any types they both share are canonicalized the same.

Reply:
Ah ok right, those are snapshots of a list at a point in time, ok makes sense!